### PR TITLE
[ADVAPP-544]: Introduce ability to duplicate forms, surveys, and events

### PR DESCRIPTION
**Pull request by dgoetzit**
### Ticket(s) or GitHub Issue

- https://canyongbs.atlassian.net/browse/ADVAPP-544

### Technical Description

> Adds replication functionality for forms, surveys, and events.

### Any deployment steps required?

> No

_______________________________________________

### CODE REVIEW

File: app-modules/form/src/Models/Form.php [hunk 1]
```diff
@@ -88,4 +90,78 @@ public function emailAutoReply(): HasOne
     {
         return $this->hasOne(FormEmailAutoReply::class);
     }
+
+    public function replicateRelatedData(Model $original): void
```
Comment (Orrison) [inline app-modules/form/src/Models/Form.php:94]: If possible, I would prefer we move this and the related methods to an Action

File: app-modules/form/tests/Filament/Resources/FormResource/Pages/ListFormsTest.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    // And we duplicate it
+    livewire(ListForms::class)
+        ->assertStatus(200)
+        ->callTableAction('Duplicate', $form);
+
+    // The form, along with all of its content, should be duplicated
+    expect(Form::count())->toBe(2);
+    expect(Form::where('id', '<>', $form->id)->first()->name)->toBe("Copy - {$form->name}");
+    expect(Form::where('id', '<>', $form->id)->first()->fields->count())->toBe($form->fields->count());
+    expect(Form::where('id', '<>', $form->id)->first()->steps->count())->toBe($form->steps->count());
```
Comment (Orrison) [inline app-modules/form/tests/Filament/Resources/FormResource/Pages/ListFormsTest.php:62]: Let's also test that no submissions were duplicated.

File: app-modules/meeting-center/src/Models/Event.php [hunk 3]
```diff
@@ -71,4 +73,78 @@ public function attendees(): HasMany
     {
         return $this->hasMany(EventAttendee::class, 'event_id');
     }
+
+    public function replicateRelatedData(Model $original): void
```
Comment (Orrison) [inline app-modules/meeting-center/src/Models/Event.php:77]: Same as Form, I would prefer this be moved to a separate Action class

File: app-modules/survey/src/Models/Survey.php [hunk 4]
```diff
@@ -82,4 +84,70 @@ public function submissions(): HasMany
     {
         return $this->hasMany(SurveySubmission::class);
     }
+
+    public function replicateRelatedData(Model $original): void
```
Comment (Orrison) [inline app-modules/survey/src/Models/Survey.php:88]: Same as Form, I would prefer this be moved to a separate Action class

File: app-modules/survey/tests/Filament/Resources/SurveyResource/Pages/ListSurveysTest.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    // And we duplicate it
+    livewire(ListSurveys::class)
+        ->assertStatus(200)
+        ->callTableAction('Duplicate', $survey);
+
+    // The survey, along with all of its content, should be duplicated
+    expect(Survey::count())->toBe(2);
+    expect(Survey::where('id', '<>', $survey->id)->first()->name)->toBe("Copy - {$survey->name}");
+    expect(Survey::where('id', '<>', $survey->id)->first()->fields->count())->toBe($survey->fields->count());
+    expect(Survey::where('id', '<>', $survey->id)->first()->steps->count())->toBe($survey->steps->count());
```
Comment (Orrison) [inline app-modules/survey/tests/Filament/Resources/SurveyResource/Pages/ListSurveysTest.php:60]: Let's also test the submissions are not duplicated

**Review by Orrison — CHANGES_REQUESTED** [review]
Let's also add a ListEventsTest similiar to the other two you wrote.

File: app-modules/survey/tests/Filament/Resources/SurveyResource/Pages/ListSurveysTest.php [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    expect($duplicatedSurvey->steps->count())->toBe($survey->steps->count());
+});
+
+it('will not duplicate survey submissions if they exist', function () {
+    asSuperAdmin();
+
+    // Given that we have a survey
+    $survey = Survey::factory()->create();
+
+    $submissionCount = $survey->submissions()->count();
```
Comment (Orrison) [inline app-modules/survey/tests/Filament/Resources/SurveyResource/Pages/ListSurveysTest.php:73]: Same as above

**Review by Orrison — CHANGES_REQUESTED** [review]

File: app-modules/form/tests/Filament/Resources/FormResource/Pages/ListFormsTest.php [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+    $submissionCount = $form->submissions()->count();
+
+    // And we duplicate it
+    livewire(ListForms::class)
+        ->assertStatus(200)
+        ->callTableAction('Duplicate', $form);
+
+    // The form submissions should not be duplicated
+    expect(FormSubmission::count())->toBe($submissionCount);
```
[the remark this answers is not in the document]
Comment (dgoetzit) [reply app-modules/form/tests/Filament/Resources/FormResource/Pages/ListFormsTest.php:83]: Yep, the factory will create between 1 and 3 submissions by default.

File: app-modules/survey/tests/Filament/Resources/SurveyResource/Pages/ListSurveysTest.php (hunk 6 above)
Comment (dgoetzit) [reply app-modules/survey/tests/Filament/Resources/SurveyResource/Pages/ListSurveysTest.php:73]: Same as forms.

File: app-modules/meeting-center/tests/Filament/Resources/EventResource/ListEventsTest.php [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    expect($duplicatedEvent->eventRegistrationForm->steps->count())->toBe($event->eventRegistrationForm->steps->count());
+});
+
+it('will not duplicate event registration form submissions if they exist', function () {
+    asSuperAdmin();
+
+    // Given that we have an event
+    $event = Event::factory()->create();
+
+    $submissionCount = $event->eventRegistrationForm->submissions()->count();
```
[the remark this answers is not in the document]
Comment (dgoetzit) [reply app-modules/meeting-center/tests/Filament/Resources/EventResource/ListEventsTest.php:76]: Same as forms.

**Review by Orrison — APPROVED** [review]

**Review by Orrison — CHANGES_REQUESTED** [review]
This looks good but there seems to be some test failure.

**Comment by Orrison** [comment]
@mxm1070 and @dgoetzit. I see that @mxm1070 re-ran the tests in CI manually, they passed and so he re-requested my review.

I ran them again manually and I am seeing the same errors pop up again. So there is something going on / some kind of flakiness here.

**Comment by mxm1070** [comment]
@Orrison Yeah I think I've seen this one before so it's not tied to this pr. I believe it's tied to an option I added awhile ago to turn off console auditing because it was creating too many audits when I was seeding while working on the api. The config should default to true in absence of the `AUDIT_CONSOLE` env key, but I suspect CI doesn't get the the config value for some reason. I remember I had to make an `.env.testing` locally and set `AUDIT_CONSOLE=true` to fix it.

**Review by Orrison — APPROVED** [review]

**Pull request merged**